File: 75. Sort Colors/75. Sort Colors.rb
# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def sort_colors(nums)
    low, mid, high = 0, 0, nums.length - 1
    
    while mid <= high
        case nums[mid]
        when 0
            nums[low], nums[mid] = nums[mid], nums[low]
            low += 1
            mid += 1
        when 1
            mid += 1
        when 2
            nums[mid], nums[high] = nums[high], nums[mid]
            high -= 1
        end
    end
end

File: 75. Sort Colors/75. Sort Colors.c
void sortColors(int* nums, int numsSize) {
    int low = 0, mid = 0, high = numsSize - 1;
    
    while (mid <= high) {
        switch (nums[mid]) {
            case 0:
                {
                    int temp = nums[low];
                    nums[low] = nums[mid];
                    nums[mid] = temp;
                    low++;
                    mid++;
                }
                break;
            case 1:
                mid++;
                break;
            case 2:
                {
                    int temp = nums[high];
                    nums[high] = nums[mid];
                    nums[mid] = temp;
                    high--;
                }
                break;
        }
    }
}